Examples/Algorithms/Geant4HepMC/src/EventAction.cpp

```diff
@@ -138,6 +142,39 @@ void followOutgoingParticles(HepMC3::GenEvent& event,
     followOutgoingParticles(event, particle->end_vertex(), processFilter);
   }
 }
+  
+HepMC3::GenParticlePtr findLastParticleTrivialAncestor( HepMC3::GenParticlePtr particle,const std::vector<std::string>& processFilter) {
+  auto endvtx=particle->end_vertex();
+  if (!endvtx) return particle;
+  if (endvtx->particles_in().size()!=1) return particle;
+  if (endvtx->particles_out().size()!=1) return particle;
+  if (!findAttribute(endvtx,processFilter)) return particle;
+  return  findLastParticleTrivialAncestor(endvtx->particles_out()[0], processFilter);
+}
+
+void followOutgoingParticlesTrivialAncestor(HepMC3::GenEvent& event, const std::vector<std::string>& processFilter) {
+   int nupdated=0;    
+   for (;;){
+     nupdated=0;
```

Comment:
Could you explain this loop and the break conditions? It looks to me that you break the inner loop as soon as you filtered once. The `nupdated == 0` condition is also not clear to me.

Reply:
The `nupdated` is the number of changes required and made in the event during the checks in the `for` loop. If no changes are made, the event is clear to go.  Because of the `break` at the end of the `for` loop, the possible values are either `0` or `1`.  Thus, the cleanup works as 'make one change to the event per pass if required and reconsider the event'. This is the simplest and safest approach, but of course, it can be improved.

Reply:
Ah okay, I see. Wouldn't a `bool` do the same job? Since it dosn't acutally count the number of updates, the variable name is a bit misleading.

Reply:
Yes, bool would do the job in this case. But the variable was made `int` for a better debug and possible optimization so the name should not be misleading.

Reply:
This code is not used in any runtime critical paths, so reliable results are more important than speed. From that point I'm fine with the current workflow. What could you debug with this variable as `int` that wouldn't be possible with a `bool`?

Reply:
Output a number of updates in one pass. But yes, I agree that is not super important.

Reply:
Ok. Not it is bool closing this thread.